dp/generate_prsn.py

```diff
@@ -0,0 +1,272 @@
+        Unit('kg / m**2 / d')
+    ]
+    pr_variable = pr.variables['pr']
+    pr_units = Unit.from_udunits_str(pr_variable.units)
+
+    if pr_units not in valid_units:
+        logger.warning('Unexpected precipitation units {}'.format(pr_units))
+        return False
+    else:
+        return True
+
+
+def preprocess_checks(pr, tasmin, tasmax, variables, required_vars):
+    '''Perform all pre-processing checks'''
+    return matching_datasets([pr, tasmin, tasmax]) and \
+        has_required_vars([pr, tasmin, tasmax], required_vars) and \
+        matching_temperature_units(tasmin, tasmax) and \
+        check_pr_units(pr) and \
+        unique_shape(variables)
+
+
+def process_to_prsn(pr, tasmin, tasmax, output_dataset, freezing):
+    '''Process precipitation data into snowfall data
+
+       This method takes a precipitation, tasmin and tasmax file and uses them
+       to produce a snowfall file.  Using a mean of the temperature files the
+       script will look through all the cells in the datacube and 0 any cells
+       that are not freezing.  This mask applies to the precipiation file and
+       thus we are left with cells where it was cold enough for snow.
+
+       The input files are read in small chunks to avoid filling up all
+       available memory and crashing the script.  Size refers to the chunk size
+       being read in from the time dimension.  Example: Chunk read in
+       (100, all lon, all lat).
+
+       Parameters:
+            pr (Variable): Variable object for precipitation
+            tasmin (Variable): Variable object for tasmin
+            tasmax (Variable): Variable object for tasmax
+            output_dataset (CFDataset): Dataset for prsn output
+            freezing (float): Freezing temperature
+    '''
+    # TODO: Figure out a way to dynamically choose chunksize
+    size = 100
+    start = 0
+    max_len = len(pr)
+    end = size
+
+    if size > max_len:
+        size = max_len
+
+    while(start < max_len):
+        if end > max_len:
+            end = max_len
+
+        pr_data = pr[start:end]
+        tasmin_data = tasmin[start:end]
+        tasmax_data = tasmax[start:end]
+
+        means = np.mean([tasmin_data, tasmax_data], axis=0)
+        prsn_data = np.where(means < freezing, pr_data, 0)
+        output_dataset.variables['prsn'][start:end] = prsn_data
+
+        start = end
+        end += size
+
+
+def generate_prsn_file(pr_filepath, tasmin_filepath, tasmax_filepath, outdir):
+    '''Generate precipiation as snow data using pr, tasmin and tasmax.
+
+       Parameters:
+            pr_filepath (str): The filepath to desired precipiation data
+            tasmin_filepath (str): The filepath to desired tasmin data
+            tasmax_filepath (str): The filepath to desired tasmax data
+            outdir (str): Output directory
+    '''
+    for filepath in [pr_filepath, tasmin_filepath, tasmax_filepath]:
+        logger.info('Retrieving file: {}'.format(filepath))
+
+    # datasets
```

Comment:
How about changing the filepaths parameter to be a hash. Then you can do something like:

```
datasets = { varname: CFDataset(filepath) for varname, filepath in filepaths.items() }
variables = { varname: datasets[varname].variables[varname] for varname in filepaths.keys() }
required_vars = filepaths.keys()
```

Simplifies a bit of the repetition maybe?